activation/atx_service_db_test.go

```
@@ -0,0 +1,104 @@
+	atxSvc := newTestDbAtxService(t)
+
+	// Invalid chain with high height
+	sigInvalid, err := signing.NewEdSigner()
+	require.NoError(t, err)
+	invalidAtx := newInitialATXv1(t, atxSvc.golden)
+	invalidAtx.Sign(sigInvalid)
+	vInvalidAtx := toAtx(t, invalidAtx)
+	vInvalidAtx.TickCount = 100
+	require.NoError(t, err)
+	require.NoError(t, atxs.Add(atxSvc.db, vInvalidAtx, invalidAtx.Blob()))
+	atxSvc.atxsdata.AddFromAtx(vInvalidAtx, false)
+
+	// Valid chain with lower height
+	sigValid, err := signing.NewEdSigner()
+	require.NoError(t, err)
+	validAtx := newInitialATXv1(t, atxSvc.golden)
+	validAtx.NumUnits += 10
+	validAtx.Sign(sigValid)
+	vValidAtx := toAtx(t, validAtx)
+	require.NoError(t, atxs.Add(atxSvc.db, vValidAtx, validAtx.Blob()))
+	atxSvc.atxsdata.AddFromAtx(vValidAtx, false)
+
+	atxSvc.validator.(*MocknipostValidator).EXPECT().
+		VerifyChain(gomock.Any(), invalidAtx.ID(), atxSvc.golden, gomock.Any()).
+		Return(errors.New("this is invalid"))
+	atxSvc.validator.(*MocknipostValidator).EXPECT().
+		VerifyChain(gomock.Any(), validAtx.ID(), atxSvc.golden, gomock.Any())
+
+	posAtxID, err := atxSvc.PositioningATX(context.Background(), validAtx.PublishEpoch)
+	require.NoError(t, err)
+	require.Equal(t, vValidAtx.ID(), posAtxID)
+
+	// look in a later epoch, it should return the same one (there is no newer one).
+	atxSvc.validator.(*MocknipostValidator).EXPECT().
+		VerifyChain(gomock.Any(), invalidAtx.ID(), atxSvc.golden, gomock.Any()).
+		Return(errors.New(""))
+	atxSvc.validator.(*MocknipostValidator).EXPECT().
+		VerifyChain(gomock.Any(), validAtx.ID(), atxSvc.golden, gomock.Any())
+
+	posAtxID, err = atxSvc.PositioningATX(context.Background(), validAtx.PublishEpoch+1)
+	require.NoError(t, err)
+	require.Equal(t, vValidAtx.ID(), posAtxID)
+
+	// it returns the golden ATX if couldn't find a better one
+	posAtxID, err = atxSvc.PositioningATX(context.Background(), validAtx.PublishEpoch-1)
+	require.NoError(t, err)
+	require.Equal(t, atxSvc.golden, posAtxID)
```

I don't think the ATX DB service even needs to know what the golden ATX is and should just return an error in this case `ErrNotFound`?

---

It needs to know the golden ATX for verifying the positioning ATX chain validity (it needs to know when to stop on a golden ATX). 

However, it could return `ErrNotFound` instead of falling back to the golden ATX and let the caller decide what to do.

---

> However, it could return ErrNotFound instead of falling back to the golden ATX and let the caller decide what to do.

Sounds sensible to me 🙂